returnn/tf/layers/basic.py

```diff
@@ -852,46 +853,54 @@ class SliceNdLayer(_ConcatInputLayer):
-    self.size = size
-    start = tf.expand_dims(start, axis=1)  # (B, T)
-    slices = slice_nd(x.placeholder, start=tf.cast(start, tf.int32), size=size)  # (B,size, ...)
+    start = start.output.copy_as_batch_major()  # e.g. (B,) or (B,T)
+    # build Data object for the position argument of GatherLayer
+    indices_data = start.copy_template(name="%s_gather_indices" % self.name)
+    indices_data = indices_data.copy_add_spatial_dim(spatial_dim_axis=start.batch_ndim, auto_time_dim_axis=False,
+                                                     dim=size)
+    start = start.placeholder # e.g. (B,) or (B,T)
+    # [start+0, start+1, ...]
+    indices = tf.expand_dims(start, -1) + tf.range(0, size) # e.g. (B, size) or (B, T, size)
     if seq_lens is not None:
-      mask = tf.greater_equal(tf.range(size)[None, :] + start, seq_lens[:, None])  # (B,T)
-      mask = expand_multiple_dims(mask, list(range(2, x.batch_ndim)))
-      slices = where_bc(mask, tf.zeros_like(slices), slices)
-    size_placeholder = x.size_placeholder.copy()
-    if isinstance(size, tf.Tensor):
-      size_placeholder[0] = tf.maximum(seq_lens - tf.reshape(start, tf.shape(seq_lens)), 0)
-      tag = DimensionTag(
-        description="sliced-time:%s" % self.get_absolute_name(),
-        kind=DimensionTag.Types.Spatial, batch=self.output.batch)
-      tag.set_tag_on_size_tensor(size_placeholder[0])
+      # broadcast from (B,) to the shape of the indices
+      seq_lens = expand_multiple_dims(seq_lens, [-1] * len(indices.shape[1:])) # e.g. (B,1) or (B,1,1)
+      # mask if an index is larger than the length of a sequence
+      pad_mask = tf.logical_or(tf.greater(indices, seq_lens-1), tf.less(indices, 0)) # shape like indices
+      # clip indices to the min and max indices of the sequences so that GatherLayer does not fail
+      indices = tf.clip_by_value(indices, 0, seq_lens-1)
     else:
-      assert isinstance(size, int)
-      size_placeholder.pop(0, None)  # static time axis
-    self.output.size_placeholder = size_placeholder
-    self.output.placeholder = slices
+      # mask if an index is larger than the length of a sequence
+      pad_mask = tf.logical_or(tf.greater(indices, x.batch_shape[1] - 1), tf.less(indices, 0))
+      # clip indices to the min and max indices of the sequences
+      indices = tf.clip_by_value(indices, 0, x.batch_shape[1] - 1)
+    indices_data.placeholder = indices
+    position = InternalLayer(
+      network=self.network,
+      name="%s_internal" % indices_data.name,
+      output=indices_data
+    )
+    gather_layer = GatherLayer(
+      name="%s_gather" % self.name,
+      network=self.network,
+      output=self.output,
+      sources=self.sources,
+      position=position,
+      axis="T"
+    )
+    placeholder = gather_layer.output.placeholder
+    self.output.size_placeholder = gather_layer.output.size_placeholder
+    # zero padding
+    self.output.placeholder = where_bc(pad_mask, tf.zeros_like(placeholder), placeholder)
```

Comment:
I don't understand this masking. The pad mask seems to match for the input sources but not for the gather layer output? And in any case, why do you need it? You should never do extra manual padding for no real reason.

Reply:
The `where_bc` function also matches the dimensions of the gather layer output, but I can also do this explicitly. The padding mask is needed because I clip the gather indices to the lowest and highest possible values of the sequences. Therefore, the GatherLayer gathers the same values from the input for the clipped indices. These values then need to be zero-padded. In the previous implementation of the SliceNdLayer, the same is done inside the slice_nd function.

Reply:
> These values then need to be zero-padded.

But why?

Reply:
> But why?

I thought this would be the desired behavior because it was done in the previous implementation. What would you suggest we do instead? Couldn't it break existing setups if we change this behavior?

Reply:
Normally, anything outside the padding should not really matter at all.

Strictly speaking, this is not always exactly true. E.g. there is `ConvLayer` which would use the padded values.

Or other layers could make use of the padding if it is the right one (via `get_padding_info_dict_ref`).

However, often the user might actually not care too much and ignore the padding (even if incorrect), or use this small incorrectness as a compromise for better speed.

It's not so nice that we mostly ignore this, although the number of cases where it makes a (small) difference is rare (e.g. `ConvLayer`), and in cases where it is important (e.g. `SoftmaxOverSpatialLayer`, `ReduceLayer`), we will anyway apply some masking.

It's also not nice that we have this somewhat inconsistent. E.g. here it is used in the previous `SliceNdLayer` implementation but not in most other layers.

However, it is safe to leave it in. It will just give you a slight inefficiency.

I wonder though why I put it there in the first place. Probably I had some reason, but I don't remember.

Depending on follup layers, you might get some gradient for these frames as well. And due to the gathering, maybe it became very wrong then.

Maybe @Zettelkasten also has a comment here.

Reply:
This kind of looks like it reproduces the behavior how invalid indices are gathered in TF. On GPU, `tf.gather` will output zero for invalid indices. Our implementation here would return the first pos if we wouldn't mask it away afterwards.
So maybe this masking is just to be consistent with the Tensorflow behavior.

But I also don't think this is necessary. Like you said.

Reply:
Let's just leave it in, but add some comment explaining the possible gradient issue.

Reply:
Okay, what comment should I add here? Just something like: "This could lead to some gradient issue if the padded frames are used by follow-up layers"?

Reply:
Leave away the "if the padded frames are used by follow-up layers". It's always used. And it can end up in the gradient in unexpected ways. Write sth like:
```
# In principle, the padded frames are being ignored
# (unless get_padding_info_dict_ref et al are used).
# However, you can still end up with gradients for them
# in unexpected ways.
# Due to our gather implementation,
# the gradient flow would go into wrong frames
# and might lead to unexpected behavior.
# So to be on the safe side, we do the masking here.
```